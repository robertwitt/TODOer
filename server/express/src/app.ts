import express from "express";
import YAML from "yamljs";
import path from "path";
import { connector } from "swagger-routes-express";
import * as apiHandlers from "./api";
import { handleError } from "./middleware/error";

const app = express();

app.set("port", process.env.PORT || 3000);

// Setup body parsers
app.use(express.json());
app.use(express.text());
app.use(express.urlencoded({ extended: false }));

// Connect API and app
const apiDefinition = YAML.load(path.join("resources", "api.yaml"));
const connect = connector(apiHandlers, apiDefinition);
connect(app);

// Add error handler
app.use(handleError);

export default app;
