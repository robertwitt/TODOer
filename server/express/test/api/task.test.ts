import app from "../../src/app";
import supertest from "supertest";
import MockDb from "../../src/repository/mock/db";

describe("The API server", () => {
  const request = supertest(app);

  beforeEach(() => MockDb.instance.reset());

  it("can read a single task by ID", async () => {
    const { body } = await request.get("/beta/Tasks/1").expect(200);
    expect(body).toMatchSnapshot();
  });

  it("cannot read a task by an unknown ID", async () => {
    await request.get("/beta/Tasks/99").expect(404);
  });

  it("can read all tasks", async () => {
    const { body } = await request.get("/beta/Tasks").expect(200);
    expect(body).toEqual({
      value: expect.arrayContaining([expect.objectContaining({ id: 1 })]),
    });
  });

  it("can filter tasks by collection", async () => {
    const { body } = await request.get("/beta/Tasks?collection=42").expect(200);
    expect(body).toMatchSnapshot();
  });

  it("can create a new task", async () => {
    const { body } = await request
      .post("/beta/Tasks")
      .send({
        collection: 1,
        title: "new task",
        dueDate: "2021-04-08",
        isPlannedForMyDay: true,
      })
      .expect(201);
    expect(body).toMatchSnapshot();
    await request.get(`/beta/Tasks/${body.id}`).expect(200);
  });

  it("fails creating a task with malformed due date", async () => {
    await request
      .post("/beta/Tasks")
      .send({ collection: 1, dueDate: "20210408" })
      .expect(400);
  });

  it("fails creating a task without collection", async () => {
    await request.post("/beta/Tasks").send({ title: "new task" }).expect(400);
  });

  it("fails creating a task for a task list view", async () => {
    await request.post("/beta/Tasks").send({ title: "new task", collection: 2 }).expect(400);
  });

  it("can update a task", async () => {
    const { body } = await request
      .patch("/beta/Tasks/1")
      .send({
        dueDate: "2021-04-10",
        dueTime: "18:00:00",
        isPlannedForMyDay: true,
      })
      .expect(200);
    expect(body).toMatchSnapshot();
  });

  it("cannot update a done task", async () => {
    await request
      .patch("/beta/Tasks/42")
      .send({ title: "updated" })
      .expect(400);
  });

  it("fails updating a task with unknown collection", async () => {
    await request.patch("/beta/Tasks/1").send({ collection: 99 }).expect(400);
  });

  it("fails updating a task with nulled collection", async () => {
    await request.patch("/beta/Tasks/1").send({ collection: null }).expect(400);
  });

  it("can delete a task by its ID", async () => {
    await request.delete("/beta/Tasks/1").expect(204);
    await request.get("/beta/Tasks/1").expect(404);
  });

  it("cannot delete a done task", async () => {
    await request.delete("/beta/Tasks/42").expect(400);
    await request.get("/beta/Tasks/42").expect(200);
  });

  it("can set a task to done", async () => {
    await request.post("/beta/Tasks/1/setToDone").expect(204);
    const { body } = await request.get("/beta/Tasks/1").expect(200);
    expect(body).toMatchSnapshot();
  });

  it("can cancel a task", async () => {
    await request.post("/beta/Tasks/1/cancel").expect(204);
    const { body } = await request.get("/beta/Tasks/1").expect(200);
    expect(body).toMatchSnapshot();
  });

  it("can reopen a task", async () => {
    await request.post("/beta/Tasks/42/reopen").expect(204);
    const { body } = await request.get("/beta/Tasks/42").expect(200);
    expect(body).toMatchSnapshot();
  });

  it("can update status with idempotent calls", async () => {
    await request.post("/beta/Tasks/42/setToDone").expect(204);
    const { body } = await request.get("/beta/Tasks/42").expect(200);
    expect(body).toMatchSnapshot();
  });
});
