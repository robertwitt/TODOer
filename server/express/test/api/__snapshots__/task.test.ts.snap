// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The API server can cancel a task 1`] = `
Object {
  "collection": Object {
    "id": 1,
    "title": "My Tasks",
  },
  "dueDate": "2021-05-31",
  "dueTime": null,
  "id": 1,
  "isDeletable": false,
  "isPlannedForMyDay": false,
  "isUpdatable": false,
  "priority": Object {
    "code": 5,
    "name": "low",
  },
  "status": Object {
    "code": "X",
    "name": "cancelled",
  },
  "title": "Tax declaration",
}
`;

exports[`The API server can create a new task 1`] = `
Object {
  "collection": Object {
    "id": 1,
    "title": "My Tasks",
  },
  "dueDate": "2021-04-08",
  "dueTime": null,
  "id": 43,
  "isDeletable": true,
  "isPlannedForMyDay": true,
  "isUpdatable": true,
  "priority": null,
  "status": Object {
    "code": "O",
    "name": "open",
  },
  "title": "new task",
}
`;

exports[`The API server can filter tasks by collection 1`] = `
Object {
  "value": Array [
    Object {
      "collection": Object {
        "id": 42,
        "title": null,
      },
      "dueDate": null,
      "dueTime": null,
      "id": 42,
      "isDeletable": false,
      "isPlannedForMyDay": false,
      "isUpdatable": false,
      "priority": null,
      "status": Object {
        "code": "D",
        "name": "done",
      },
      "title": "What is the meaning of life?",
    },
  ],
}
`;

exports[`The API server can read a single task by ID 1`] = `
Object {
  "collection": Object {
    "id": 1,
    "title": "My Tasks",
  },
  "dueDate": "2021-05-31",
  "dueTime": null,
  "id": 1,
  "isDeletable": true,
  "isPlannedForMyDay": false,
  "isUpdatable": true,
  "priority": Object {
    "code": 5,
    "name": "low",
  },
  "status": Object {
    "code": "O",
    "name": "open",
  },
  "title": "Tax declaration",
}
`;

exports[`The API server can reopen a task 1`] = `
Object {
  "collection": Object {
    "id": 42,
    "title": null,
  },
  "dueDate": null,
  "dueTime": null,
  "id": 42,
  "isDeletable": true,
  "isPlannedForMyDay": false,
  "isUpdatable": true,
  "priority": null,
  "status": Object {
    "code": "O",
    "name": "open",
  },
  "title": "What is the meaning of life?",
}
`;

exports[`The API server can set a task to done 1`] = `
Object {
  "collection": Object {
    "id": 1,
    "title": "My Tasks",
  },
  "dueDate": "2021-05-31",
  "dueTime": null,
  "id": 1,
  "isDeletable": false,
  "isPlannedForMyDay": false,
  "isUpdatable": false,
  "priority": Object {
    "code": 5,
    "name": "low",
  },
  "status": Object {
    "code": "D",
    "name": "done",
  },
  "title": "Tax declaration",
}
`;

exports[`The API server can update a task 1`] = `
Object {
  "collection": Object {
    "id": 1,
    "title": "My Tasks",
  },
  "dueDate": "2021-04-10",
  "dueTime": "18:00:00",
  "id": 1,
  "isDeletable": true,
  "isPlannedForMyDay": true,
  "isUpdatable": true,
  "priority": Object {
    "code": 5,
    "name": "low",
  },
  "status": Object {
    "code": "O",
    "name": "open",
  },
  "title": "Tax declaration",
}
`;

exports[`The API server can update status with idempotent calls 1`] = `
Object {
  "collection": Object {
    "id": 42,
    "title": null,
  },
  "dueDate": null,
  "dueTime": null,
  "id": 42,
  "isDeletable": false,
  "isPlannedForMyDay": false,
  "isUpdatable": false,
  "priority": null,
  "status": Object {
    "code": "D",
    "name": "done",
  },
  "title": "What is the meaning of life?",
}
`;
